chaosreliably/controls/experiment.py

```diff
@@ -0,0 +1,310 @@
+from typing import Any, cast
+
+from chaoslib.types import Configuration, Experiment, Secrets
+from logzero import logger
+
+from chaosreliably import get_session
+from chaosreliably.types import (
+    EntityContext,
+    EntityContextExperimentEventLabels,
+    EntityContextExperimentLabels,
+    EntityContextExperimentRunLabels,
+    EntityContextExperimentVersionLabels,
+    EntityContextMetadata,
+    EventType,
+)
+
+__all__ = ["before_experiment_control"]
+
+
+def before_experiment_control(
+    context: Experiment,
+    configuration: Configuration = None,
+    secrets: Secrets = None,
+    **kwargs: Any,
+) -> None:
+    """
+    Control run *before* the execution of an Experiment.
+
+    For a given Experiment, the control creates (if not already created) an Experiment
+    Entity Context and an Experiment Version Entity Context in the Reliably Entity
+    Server.
+
+    A unique Experiment Run Entity Context is also created, with an Experiment Event
+    Entity Context of type `EXPERIMENT_START` created, relating to the run.
+
+    The control requires the `arguments` of `commit_hash`, `source`, and `user` to be
+    provided to the control definition. If not provided, the control will simply not
+    create any Entity Contexts.
+
+    Once the Entity Contexts have been created, an entry into the configuration is made
+    under configuration["chaosreliably"]["experiment_run_labels"] to allow for
+    other controls to create events relating to the Experiment Run.
+
+    :param context: Experiment object representing the Experiment that will be executed
+    :param configuration: Configuration object provided by Chaos Toolkit
+    :param secrets: Secret object provided by Chaos Toolkit
+    :param **kwargs: Expected `kwargs` are 'commit_hash' (str), `source` (str), and
+        `user` (str)
+
+    Examples
+    --------
+
+    "controls": [
+        {
+            "name": "chaosreliably",
+            "provider": {
+                "type": "python",
+                "module": "chaosreliably.controls"
+                "arguments": {
+                    "commit_hash": "59f9f577e2d90719098f4d23d26329ce41f2d0bd",
+                    "source": "https://github.com/chaostoolkit-incubator/chaostoolkit-reliably/exp.json",  # Noqa
+                    "user": "A users name"
+                }
+            }
+        }
+    ]
+    """
+    try:
+        commit_hash = kwargs.get("commit_hash")
+        source = kwargs.get("source")
+        user = kwargs.get("user")
+        if not commit_hash or not source or not user:
+            logger.warn(
```

Comment:
Cool, log is here. 

`logger.warn` is deprecated however https://docs.python.org/3/library/logging.html#logging.Logger.warning

Also, I'm not sure warning level makes sense since controls are usually transparent to the experiment side. I would use a debug.

Reply:
We're using `logzero` so afaik `.warn` is still valid in that, but I can switch to debug anyway

Reply:
the logger from logzero is just a Python logger. logzero merely makes the creation simpler IIRC.